refactor bindX & fix bindV

vicb opened the pull request:
Easier to understand.

vicb committed:
refactor(ivy): simplify bind code

mary-poppins commented:
You can preview ffd07c9 at https://pr21881-ffd07c9.ngbuilds.io/.

mary-poppins commented:
You can preview 6dde82c at https://pr21881-6dde82c.ngbuilds.io/.

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1446,22 +1454,12 @@ export function bind1(prefix: string, value: any, suffix: string): string|NO_CHA
- *
- * @param prefix
- * @param v0 value checked for change
- * @param i0
```
Why did you remove the doc?

vicb replied:
Because I don't see any added value in 
```
- * @param prefix
 - * @param v0
 - * @param i0
 - * @param v1
 - * @param i1
 - * @param v2
 - * @param i2
 - * @param v3
 - * @param i3
 - * @param v4
 - * @param suffix
```

mary-poppins commented:
You can preview dd3ac1e at https://pr21881-dd3ac1e.ngbuilds.io/.

mary-poppins commented:
You can preview e4cadad at https://pr21881-e4cadad.ngbuilds.io/.

vicb committed:
fix(ivy): improve `bindV` perf and memory usage

- Fix the case when first dynamic values are NO_CHANGE
- Do not store the static texts (even indexes) as bindings,
- Do not diff static texts (they do not change),
- Do not stringify static texts,
- Remove superfluous values walking.

mary-poppins commented:
You can preview d572eed at https://pr21881-d572eed.ngbuilds.io/.

mary-poppins commented:
You can preview e1158ad at https://pr21881-e1158ad.ngbuilds.io/.

chuckjaz commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1372,33 +1372,38 @@ export const NO_CHANGE = {} as NO_CHANGE;
+ * - has evaluated expressions at odd indexes (could be NO_CHANGE).
  */
 export function bindV(values: any[]): string|NO_CHANGE {
+  ngDevMode && assertLessThan(2, values.length, 'should have at least 3 values');
```
Should we assert that it has at least 9 values as the more optimal binds should be called instead?

vicb replied:
The code works with 3+ values but I agree it should not be called with less than 9.
We should probably have a mechanism to check that (out of scope of this PR)

mary-poppins commented:
You can preview f9f6e1b at https://pr21881-f9f6e1b.ngbuilds.io/.
